### PR TITLE
ADEN-9025 Adding HideOnViewability template timeout option

### PR DESCRIPTION
**Pull request by dmnsobczak**

### CODE REVIEW

File: src/ad-products/templates/hide-on-viewability.ts [hunk 1]
```diff
@@ -47,24 +49,46 @@ export class HideOnViewability {
...
+		}
+
+		Promise.race([this.registerViewabilityListener(this.adSlot), this.registerTimeoutHide()]).then(
+			() => {
+				this.hideSlot(this.adSlot);
+			},
+		);
+	}
+
+	private registerViewabilityListener(adSlot: AdSlot): Promise<void> {
```
Comment (fraszczakszymon) [inline src/ad-products/templates/hide-on-viewability.ts]: s/registerViewabilityListener/getViewabilityPromise

File: src/ad-products/templates/hide-on-viewability.ts [hunk 2]
```diff
@@ -47,24 +49,46 @@ export class HideOnViewability {
...
+				}, this.config.additionalHideTime || 0);
+			});
+		});
 	}
 
-	private registerViewabilityListener(adSlot: AdSlot): void {
-		adSlot.on(AdSlot.SLOT_VIEWED_EVENT, () => {
-			if (!this.config || !this.config.hideAnimation) {
-				return;
+	private registerTimeoutHide(): Promise<void> {
```
Comment (fraszczakszymon) [inline src/ad-products/templates/hide-on-viewability.ts]:
```suggestion
	private getTimeoutHidePromise(): Promise<void> {
```

File: src/ad-products/templates/hide-on-viewability.ts [hunk 3]
```diff
@@ -47,24 +49,46 @@ export class HideOnViewability {
...
+
+		Promise.race([this.registerViewabilityListener(this.adSlot), this.registerTimeoutHide()]).then(
+			() => {
+				this.hideSlot(this.adSlot);
+			},
+		);
+	}
+
+	private registerViewabilityListener(adSlot: AdSlot): Promise<void> {
+		return new Promise((resolve) => {
```
Comment (fraszczakszymon) [inline src/ad-products/templates/hide-on-viewability.ts]:
```suggestion
		return new Promise((resolve) => {
			adSlot.viewed.then(() => {
				setTimeout(() => {
					utils.logger(HideOnViewability.getName(), 'viewability Promise resolved');
					resolve();
				}, this.config.additionalHideTime || 0);
			});
		});
```

Would it work? What do you think? Worth changing? Your call - I'm fine with leaving it as it is too.

File: src/ad-products/templates/hide-on-viewability.ts (hunk 2 above)
Comment (dmnsobczak) [reply src/ad-products/templates/hide-on-viewability.ts]: Done!

File: src/ad-products/templates/hide-on-viewability.ts (hunk 1 above)
Comment (dmnsobczak) [reply src/ad-products/templates/hide-on-viewability.ts]: Done!

File: src/ad-products/templates/hide-on-viewability.ts (hunk 3 above)
Comment (dmnsobczak) [reply src/ad-products/templates/hide-on-viewability.ts]: Yeah, why not. It's still working and it's more readable. Promises FTW!

File: src/ad-products/templates/hide-on-viewability.ts [hunk 4]
```diff
@@ -47,24 +49,46 @@ export class HideOnViewability {
...
+		});
 	}
 
-	private registerViewabilityListener(adSlot: AdSlot): void {
-		adSlot.on(AdSlot.SLOT_VIEWED_EVENT, () => {
-			if (!this.config || !this.config.hideAnimation) {
-				return;
+	private getTimeoutHidePromise(): Promise<void> {
+		return new Promise((resolve) => {
+			if (this.config.timeoutHideTime) {
```
Comment (fraszczakszymon) [inline src/ad-products/templates/hide-on-viewability.ts]: 1. This promise never resolves when timeout is not set. What about returning null or rejecting?
2. It does not work with falsy values (0 timeout).

Comment (dmnsobczak) [reply src/ad-products/templates/hide-on-viewability.ts]: You right. Changed.

**Pull request merged**